plan/expression_rewriter.go

```diff
@@ -19,13 +19,25 @@ import (
 // EvalSubquery evaluates incorrelated subqueries once.
 var EvalSubquery func(p PhysicalPlan, is infoschema.InfoSchema, ctx context.Context) ([]types.Datum, error)
 
-func (b *planBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int) (
+// rewrite function rewrites ast expr to expression.Expression.
+// aggMapper maps ast.AggregateFuncExpr to the columns offset in p's output schema.
+// asScalar means whether this expression must be treated as a scalar expression.
+func (b *planBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool) (
 	newExpr expression.Expression, newPlan LogicalPlan, correlated bool, err error) {
-	er := &expressionRewriter{p: p, aggrMap: aggMapper, schema: p.GetSchema(), b: b}
+	er := &expressionRewriter{
+		p:        p,
+		aggrMap:  aggMapper,
+		schema:   p.GetSchema(),
+		b:        b,
+		asScalar: asScalar,
+	}
 	expr.Accept(er)
 	if er.err != nil {
 		return nil, nil, false, errors.Trace(er.err)
 	}
+	if !er.asScalar && len(er.ctxStack) == 0 {
```

Comment:
We can use argument 'asScalar' directly as `er.asScalar` may be changed during visiting.